analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Argument.cs

```diff
@@ -36,25 +39,21 @@ private static ProgramState ProcessArgument(ProgramState state, IArgumentOperati
         {
             return null; // __arglist is not assigned to a parameter
         }
-        if (argument is { Parameter.RefKind: RefKind.Out or RefKind.Ref } && argument.Value.TrackedSymbol() is { } symbol)
+        if (argument is { Parameter.RefKind: RefKind.Out or RefKind.Ref } && argument.Value.TrackedSymbol() is { } refOutSymbol)
         {
-            state = state.SetSymbolValue(symbol, null); // Forget state for "out" or "ref" arguments
+            state = state.SetSymbolValue(refOutSymbol, null); // Forget state for "out" or "ref" arguments
         }
-        if (argument.Parameter.GetAttributes() is { Length: > 0 } attributes)
+        if (HasNotNullAttribute(argument.Parameter)
+            && argument.Value.TrackedSymbol() is { } notNullSymbol)
         {
-            state = ProcessArgumentAttributes(state, argument, attributes);
+            state = state.SetSymbolConstraint(notNullSymbol, ObjectConstraint.NotNull);
         }
         return state;
     }
 
-    private static ProgramState ProcessArgumentAttributes(ProgramState state, IArgumentOperationWrapper argument, ImmutableArray<AttributeData> attributes) =>
-        attributes.Any(IsNotNullAttribute) && argument.Value.TrackedSymbol() is { } symbol
-            ? state.SetSymbolConstraint(symbol, ObjectConstraint.NotNull)
-            : state;
-
     // https://docs.microsoft.com/dotnet/api/microsoft.validatednotnullattribute
     // https://docs.microsoft.com/dotnet/csharp/language-reference/attributes/nullable-analysis#postconditions-maybenull-and-notnull
     // https://www.jetbrains.com/help/resharper/Reference__Code_Annotation_Attributes.html#NotNullAttribute
-    private static bool IsNotNullAttribute(AttributeData attribute) =>
+    public static bool IsNotNullAttribute(AttributeData attribute) =>
         attribute.HasAnyName("ValidatedNotNullAttribute", "NotNullAttribute");
```

Comment:
Why are these `public ` now? `HasNotNullAttribute` and `IsNotNullAttribute` should be made extension methods if we want to reuse them. We will also likely add support for more of these attributes in the future. Can we add these to KnownTypes and add an overload to `AttributeSyntaxExtensions.IsKnownType` instead? (This only works if we can ignore the namespace, though).

Reply:
The `IsNotNullAttribute` as not intentional. I'll make the other one extension.